# 2020.04.06 (Mon)

## Starting the annotations from scratch (`RF sentences`)

Following the official guide:
- https://www.youtube.com/watch?v=59BKHO_xBPA
- https://github.com/explosion/projects/tree/master/ner-food-ingredients#data-creation-and-training-workflow

### Label data manually

In [2]:
!prodigy ner.manual cord_19_rf_sentences blank:en data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR

Using 1 label(s): RISK_FACTOR

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!

^C

✔ Saved 102 annotations to database SQLite
Dataset: cord_19_rf_sentences
Session ID: 2020-04-06_22-08-56



### Pretrain different models

#### `en_vectors_web_lg`

In [5]:
!prodigy train ner cord_19_rf_sentences en_vectors_web_lg --eval-split 0.2

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     389.18       0.000      0.000      0.000                                 
 2     191.03      25.000      4.000      6.897                                 
 3     265.45      35.294     24.000     28.571                                 
 4     246.57      28.571     24.000     26.087                                 
 5     239.21      22.727     20.000     21.277                                 
 6     215.80      42.857     24.000     30.769                                 
 7     258.40      42.105     32.000     36.364                                 
 8     237.19      50.000     32.000   

#### `en_core_web_lg`

In [6]:
!prodigy train ner cord_19_rf_sentences en_core_web_lg --eval-split 0.2

✔ Loaded model 'en_core_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1421.81       0.000      0.000      0.000                                 
 2    1245.16       0.000      0.000      0.000                                 
 3    1262.77       0.000      0.000      0.000                                 
 4    1249.04       0.000      0.000      0.000                                 
 5    1262.46       0.000      0.000      0.000                                 
 6    1205.26       0.000      0.000      0.000                                 
 7    1137.92       0.000      0.000      0.000                                 
 8    1131.84       0.000      0.000      

⁉️ Why is `en_core_web_lg` producing no results at all?

#### `en_core_sci_lg`

In [14]:
!prodigy train ner cord_19_rf_sentences en_core_sci_lg --eval-split 0.2 --output models/2020_04_06_rf_sentences

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     559.17       0.000      0.000      0.000                                 
 2     550.33      33.333      8.000     12.903                                 
 3     538.43      50.000     24.000     32.432                                 
 4     457.89      52.632     40.000     45.455                                 
 5     452.06      50.000     36.000     41.860                                 
 6     415.01      60.000     48.000     53.333                                 
 7     489.45      47.826     44.000     45.833                                 
 8     461.13      45.000     36.000     4

#### `en_vectors_web_lg` + tok2vec

In [8]:
!prodigy train ner cord_19_rf_sentences en_vectors_web_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_vectors_model436.bin

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_vectors_model436.bin
embed_rows: 2000 | require_vectors: True | cnn_maxout_pieces: 3 |
token_vector_width: 96 | conv_depth: 4 | nr_feature_tokens: 3 |
pretrained_vectors: en_vectors_web_lg.vectors
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     339.88       0.000      0.000      0.000                                 
 2     199.85      50.000     16.000     24.242                                 
 3     128.26      50.000     28.000     35.897                                 
 4      76.43      42.857     24.000     30.769                                 
 5      63.28      60.000     3

#### `en_core_web_lg` + tok2vec

In [9]:
!prodigy train ner cord_19_rf_sentences en_core_web_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_vectors_model436.bin

✔ Loaded model 'en_core_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_vectors_model436.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2055.18       0.000      0.000      0.000                                 
 2    1808.48       0.000      0.000      0.000                                 
 3    1624.76       0.000      0.000      0.000                                 
 4    1443.31       0.000      0.000      0.000                                 
 5    1652.93       0.000      0.000      0.000                                 
 6    1470.22       0.000      0.000      0.000                                 
 7    1345.14       0.000      0.000      0.000

#### `en_core_sci_lg` + tok2vec

In [13]:
!prodigy train ner cord_19_rf_sentences en_core_sci_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_sci_model328.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model328.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     536.68       0.000      0.000      0.000                                 
 2     631.61       0.000      0.000      0.000                                 
 3     539.99      40.000     24.000     30.000                                 
 4     505.27      47.619     40.000     43.478                                 
 5     508.05      45.000     36.000     40.000                                 
 6     461.69      52.941     36.000     42.857                                 
 7     473.25      45.000     36.000     40.000    

🔥 Top model so far: `53.333` 🔥

### Label more data by correcting the model's predictions

In [15]:
!prodigy ner.correct cord_19_rf_sentences_correct models/2020_04_06_rf_sentences data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR --exclude cord_19_rf_sentences

Using 1 label(s): RISK_FACTOR
Added dataset cord_19_rf_sentences_correct to database SQLite.

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!

^C

✔ Saved 220 annotations to database SQLite
Dataset: cord_19_rf_sentences_correct
Session ID: 2020-04-06_23-58-48



### Train the final model

For futher training, only `en_core_sci_lg` will be used.

In [3]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1554.25      61.538     37.500     46.602                                 
 2    1346.09      64.407     59.375     61.789                                 
 3    1407.15      66.667     50.000     57.143                                 
 4    1411.73      72.727     50.000     59.259                                 
 5    1412.81      73.913     53.125     61.818                                 
 6    1319.96      73.913     53.125     61.818                                 
 7    1378.05      72.917     54.688     62.500                                 
 8    1208.99      72.727     62.500    

In [6]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model328.bin --output models/2020_04_07_rf_sentences_corrected

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model328.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1578.16      64.000     25.000     35.955                                 
 2    1439.61      55.319     40.625     46.847                                 
 3    1528.20      72.222     60.938     66.102                                 
 4    1429.18      71.154     57.812     63.793                                 
 5    1414.46      69.388     53.125     60.177                                 
 6    1344.69      75.472     62.500     68.376                                 
 7    1472.62      70.909     60.938     65.546  

In [5]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model989.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model989.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1570.28      62.500     23.438     34.091                                 
 2    1397.40      59.091     40.625     48.148                                 
 3    1485.47      70.833     53.125     60.714                                 
 4    1445.56      68.293     43.750     53.333                                 
 5    1431.31      64.000     50.000     56.140                                 
 6    1339.87      70.833     53.125     60.714                                 
 7    1387.07      72.340     53.125     61.261  

# 2020.04.09 (Thu)
## Experiments with new `tok2vec` models

In [1]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model978_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1577.68      57.895     34.375     43.137                                 
 2    1386.95      67.308     54.688     60.345                                 
 3    1455.49      76.596     56.250     64.865                                 
 4    1457.04      79.592     60.938     69.027                                 
 5    1413.30      78.431     62.500     69.565                                 
 6    1351.85      76.923     62.500     68.966                                 
 7    1304.38      78.000     60.938     68.4

In [2]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model997_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model997_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1630.03      55.556     23.438     32.967                                 
 2    1442.41      55.556     46.875     50.847                                 
 3    1429.19      65.957     48.438     55.856                                 
 4    1467.20      65.455     56.250     60.504                                 
 5    1467.20      69.643     60.938     65.000                                 
 6    1376.49      67.241     60.938     63.934                                 
 7    1324.06      73.585     60.938     66.6

In [3]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_abs_fil_sci_model975_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1587.95      62.963     26.562     37.363                                 
 2    1500.37      60.784     48.438     53.913                                 
 3    1510.36      63.492     62.500     62.992                                 
 4    1519.20      60.000     56.250     58.065                                 
 5    1521.26      65.574     62.500     64.000                                 
 6    1495.73      64.912     57.812     61.157                                 
 7    1386.50      65.517     59.375    

# 2020.04.10 (Fri)
## Experiments with new `tok2vec` models (cont.)

The Kaggle notebook training models for the full set of abstracts (i.e. `cord_19_abstracts.jsonl`) crashed, so I have only partial results and no loss metrics for them.

In [11]:
import os

from pathlib import Path

# run training for a selected models, use every skip_models-th file
skip_models = 13
abs_models_dir = Path('models/tok2vec_abs_sci_ALL_gpu')

models = sorted(os.listdir(abs_models_dir))

for idx, model_file in enumerate(models):
    if idx % skip_models == 0:
        model_path = abs_models_dir / model_file
        !prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg \
            --eval-split 0.2 --n-iter 20 --init-tok2vec $model_path

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model100.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1570.29      43.478     31.250     36.364                                 
 2    1421.36      59.574     43.750     50.450                                 
 3    1486.19      68.519     57.812     62.712                                 
 4    1477.14      70.000     54.688     61.404                                 
 5    1389.42      73.469     56.250     63.717                                 
 6    1377.87      70.000     54.688     61.404                                 
 7    1379.27      65.385     53.125    

15    1143.57      69.231     56.250     62.069                                 
16    1108.60      69.091     59.375     63.866                                 
17    1037.24      66.667     59.375     62.810                                 
18    1059.90      67.857     59.375     63.333                                 
19    1029.31      65.517     59.375     62.295                                 
20    1091.05      67.797     62.500     65.041                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      76.364   65.625    70.588


Best F-Score   70.588
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs

✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model230.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1537.49      65.714     35.938     46.465                                 
 2    1506.48      51.064     37.500     43.243                                 
 3    1436.09      68.627     54.688     60.870                                 
 4    1438.86      67.925     56.250     61.538                                 
 5    1471.17      58.730     57.812     58.268                                 
 6    1395.94      76.786     67.188     71.667                                 
 7    1329.15      71.186     65.625     68.293                                 
 8    1224.20      70.000     65.625     67.742                                 
 9    1212.04      72.222     60.938     66.102                     

17    1045.76      74.545     64.062     68.908                                 
18    1047.72      74.545     64.062     68.908                                 
19    1039.92      76.786     67.188     71.667                                 
20    1076.51      75.000     65.625     70.000                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      81.818   70.312    75.630


Best F-Score   75.630
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model321.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1483.42      63.415     40.625     49.524                                 
 2    1529.89      60.000     60.938     60.465                                 
 3    1444.98      66.154     67.188     66.667                                 
 4    1448.20      68.421     60.938     64.463                                 
 5    1388.45      67.692     68.750     68.217                                 
 6    1309.81      65.574     62.500     64.000                                 
 7    1318.09      68.852     65.625     67.200                                 
 8    1223.92      66.667     65.625     66.142                                 
 9    1219.50      68.254     67.188     67.717                                 
10    1226.45      68.182     70.312     69.231                     

### Training final models with selected `tok2vec` models

In [5]:
import os

from pathlib import Path


def train_ner_with_every_tok2vec(datasets):
    tok2vec_dir = Path('models')
    tok2vecs = sorted(os.listdir(tok2vec_dir))
    
    for tok2vec_file in tok2vecs:
        if "_sci_" not in tok2vec_file:
            continue

        tok2vec_path = tok2vec_dir / tok2vec_file
        !prodigy train ner $datasets en_core_sci_lg --eval-split 0.2 --n-iter 20 \
            --init-tok2vec $tok2vec_path

In [6]:
train_ner_with_every_tok2vec("cord_19_rf_sentences,cord_19_rf_sentences_correct")

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1587.95      62.963     26.562     37.363                                 
 2    1500.37      60.784     48.438     53.913                                 
 3    1510.36      63.492     62.500     62.992                                 
 4    1519.20      60.000     56.250     58.065                                 
 5    1521.26      65.574     62.500     64.000                                 
 6    1495.73      64.912     57.812     61.157                                 
 7    1386.50      65.517     59.375    

16    1146.76      78.431     62.500     69.565                                 
17    1045.76      74.545     64.062     68.908                                 
18    1047.72      74.545     64.062     68.908                                 
19    1039.92      76.786     67.188     71.667                                 
20    1076.51      75.000     65.625     70.000                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      81.818   70.312    75.630


Best F-Score   75.630
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model336_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨ 

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1577.68      57.895     34.375     43.137                                 
 2    1386.95      67.308     54.688     60.345                                 
 3    1455.49      76.596     56.250     64.865                                 
 4    1457.04      79.592     60.938     69.027                                 
 5    1413.30      78.431     62.500     69.565                                 
 6    1351.85      76.923     62.500     68.966                                 
 7    1304.38      78.000     60.938     68.421                                 
 8    1255.67      76.471     60.938     67.826                                 
 9    1219.66      78.431     62.500     69.565                                 
10    1203.83      74.545     64.062     68.908                     

🔥 Top model so far: `75.630` 🔥
- trained with `tok2vec_abs_sci_model308_gpu.bin`

In [15]:
#!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg \
#    --eval-split 0.2 --n-iter 20 --init-tok2vec models/tok2vec_abs_sci_model308_gpu.bin \
#    --output models/2020_04_10_en_ner_rf_sm

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model308_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1518.02      61.905     40.625     49.057                                 
 2    1389.84      56.897     51.562     54.098                                 
 3    1485.12      71.698     59.375     64.957                                 
 4    1385.46      70.175     62.500     66.116                                 
 5    1455.78      72.727     62.500     67.227                                 
 6    1404.07      78.947     70.312     74.380                                 
 7    1323.94      81.818     70.312     75.

# 2020.04.13 (Mon)
## Label more data by correcting the model's predictions (part 2)

In [ ]:
!prodigy ner.correct cord_19_rf_sentences_correct_2 models/2020_04_10_en_ner_rf_sm data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR --exclude cord_19_rf_sentences,cord_19_rf_sentences_correct

Using 1 label(s): RISK_FACTOR
Added dataset cord_19_rf_sentences_correct_2 to database SQLite.

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!



## Train final models, with 3 datasets

In [7]:
train_ner_with_every_tok2vec("cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2")

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2885.30      59.302     42.149     49.275                                 
 2    2552.55      62.353     43.802     51.456                                 
 3    2728.96      65.385     56.198     60.444                                 
 4    2487.82      72.917     57.851     64.516                                 
 5    2373.22      72.642     63.636     67.841                                 
 6    2348.38      74.074     66.116     69.869                                 
 7    2218.46      73.832     65.289   

16    1818.28      75.652     71.901     73.729                                 
17    1795.52      75.000     71.901     73.418                                 
18    1827.94      74.561     70.248     72.340                                 
19    1761.02      74.783     71.074     72.881                                 
20    1806.65      73.504     71.074     72.269                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      77.193   72.727    74.894


Best F-Score   74.894
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model336_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨

✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2835.29      68.116     38.843     49.474                                 
 2    2732.17      72.500     47.934     57.711                                 
 3    2570.86      78.000     64.463     70.588                                 
 4    2535.81      72.381     62.810     67.257                                 
 5    2355.29      78.641     66.942     72.321                                 
 6    2273.33      73.333     63.636     68.142                                 
 7    2248.22      75.472     66.116     70.485                                 
 8    2121.12      72.477     65.289     68.696                                 
 9    2055.64      72.642     63.636     67.841                          

🔥 More annotations lowered the performance; top model: `74.894` 🔥
- trained with `tok2vec_abs_sci_model308_gpu.bin`

### Check if `sci` is still the best choice

In [10]:
datasets = "cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2"

for base_model in ['en_vectors_web_lg', 'en_core_web_lg', 'en_core_sci_lg']:
    !prodigy train ner $datasets $base_model --eval-split 0.2 --n-iter 20

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1517.43      63.793     30.579     41.341                                 
 2    1373.26      68.750     45.455     54.726                                 
 3     997.71      69.048     47.934     56.585                                 
 4     903.18      71.910     52.893     60.952                                 
 5     663.74      75.269     57.851     65.421                                 
 6     690.64      74.118     52.066     61.165                                 
 7     499.83      75.862     54.545     63.462                                 
 8     439.64      73.118     56.198

`en_core_sci_lg` is still the best base model, although the margins are getting smaller.